man/nominatim.1
.TH nominatim "1" Manual
.SH NAME
nominatim
.SH SYNOPSIS
.B nominatim
[-h] [--version] {import,freeze,replication,special-phrases,add-data,index,refresh,admin,export,convert,serve,search,reverse,lookup,details,status} ...
.SH DESCRIPTION
    Command\-line tools for importing, updating, administrating and
.br
    querying the Nominatim database.
.br
    
.SH OPTIONS

.TP
\fB\-\-version\fR
Print Nominatim version and exit

.SS
\fBSub-commands\fR
.TP
\fBnominatim\fR \fI\,import\/\fR
    Create a new Nominatim database from an OSM file.
.TP
\fBnominatim\fR \fI\,freeze\/\fR
    Make database read-only.
.TP
\fBnominatim\fR \fI\,replication\/\fR
    Update the database using an online replication service.
.TP
\fBnominatim\fR \fI\,special-phrases\/\fR
    Import special phrases.
.TP
\fBnominatim\fR \fI\,add-data\/\fR
    Add additional data from a file or an online source.
.TP
\fBnominatim\fR \fI\,index\/\fR
    Reindex all new and modified data.
.TP
\fBnominatim\fR \fI\,refresh\/\fR
    Recompute auxiliary data used by the indexing process.
.TP
\fBnominatim\fR \fI\,admin\/\fR
    Analyse and maintain the database.
.TP
\fBnominatim\fR \fI\,export\/\fR
    Export places as CSV file from the database.
.TP
\fBnominatim\fR \fI\,convert\/\fR
 Convert an existing database into a different format. (EXPERIMENTAL)
.TP
\fBnominatim\fR \fI\,serve\/\fR
    Start a simple web server for serving the API.
.TP
\fBnominatim\fR \fI\,search\/\fR
    Execute a search query.
.TP
\fBnominatim\fR \fI\,reverse\/\fR
    Execute API reverse query.
.TP
\fBnominatim\fR \fI\,lookup\/\fR
    Execute API lookup query.
.TP
\fBnominatim\fR \fI\,details\/\fR
    Execute API details query.
.TP
\fBnominatim\fR \fI\,status\/\fR

.SH OPTIONS 'nominatim import'
usage: nominatim import [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                        [--osm-file FILE]
                        [--continue {import-from-file,load-data,indexing,db-postprocess}]
                        [--osm2pgsql-cache SIZE] [--reverse-only]
                        [--no-partitions] [--no-updates] [--offline]
                        [--ignore-errors] [--index-noanalyse]
                        [--prepare-database]

    Create a new Nominatim database from an OSM file.
.br

.br
    This sub\-command sets up a new Nominatim database from scratch starting
.br
    with creating a new database in Postgresql. The user running this command
.br
    needs superuser rights on the database.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-osm\-file\fR FILE
OSM file to be imported (repeat for importing multiple files)

.TP
\fB\-\-continue\fR {import\-from\-file,load\-data,indexing,db\-postprocess}
Continue an import that was interrupted

.TP
\fB\-\-osm2pgsql\-cache\fR SIZE
Size of cache to be used by osm2pgsql (in MB)

.TP
\fB\-\-reverse\-only\fR
Do not create tables and indexes for searching

.TP
\fB\-\-no\-partitions\fR
Do not partition search indices (speeds up import of single country extracts)

.TP
\fB\-\-no\-updates\fR
Do not keep tables that are only needed for updating the database later

.TP
\fB\-\-offline\fR
Do not attempt to load any additional data from the internet

.TP
\fB\-\-ignore\-errors\fR
Continue import even when errors in SQL are present

.TP
\fB\-\-index\-noanalyse\fR
Do not perform analyse operations during index (expert only)

.TP
\fB\-\-prepare\-database\fR
Create the database but do not import any data

.SH OPTIONS 'nominatim freeze'
usage: nominatim freeze [-h] [-q] [-v] [--project-dir DIR] [-j NUM]

    Make database read\-only.
.br

.br
    About half of data in the Nominatim database is kept only to be able to
.br
    keep the data up\-to\-date with new changes made in OpenStreetMap. This
.br
    command drops all this data and only keeps the part needed for geocoding
.br
    itself.
.br

.br
    This command has the same effect as the `\-\-no\-updates` option for imports.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.SH OPTIONS 'nominatim replication'
usage: nominatim replication [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                             [--init] [--no-update-functions]
                             [--check-for-updates] [--once] [--catch-up]
                             [--no-index] [--osm2pgsql-cache SIZE]
                             [--socket-timeout SOCKET_TIMEOUT]

    Update the database using an online replication service.
.br

.br
    An OSM replication service is an online service that provides regular
.br
    updates (OSM diff files) for the planet or update they provide. The OSMF
.br
    provides the primary replication service for the full planet at
.br
    https://planet.osm.org/replication/ but there are other providers of
.br
    extracts of OSM data who provide such a service as well.
.br

.br
    This sub\-command allows to set up such a replication service and download
.br
    and import updates at regular intervals. You need to call '\-\-init' once to
.br
    set up the process or whenever you change the replication configuration
.br
    parameters. Without any arguments, the sub\-command will go into a loop and
.br
    continuously apply updates as they become available. Giving `\-\-once` just
.br
    downloads and imports the next batch of updates.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-init\fR
Initialise the update process

.TP
\fB\-\-no\-update\-functions\fR
Do not update the trigger function to support differential updates (EXPERT)

.TP
\fB\-\-check\-for\-updates\fR
Check if new updates are available and exit

.TP
\fB\-\-once\fR
Download and apply updates only once. When not set, updates are continuously applied

.TP
\fB\-\-catch\-up\fR
Download and apply updates until no new data is available on the server

.TP
\fB\-\-no\-index\fR
Do not index the new data. Only usable together with \-\-once

.TP
\fB\-\-osm2pgsql\-cache\fR SIZE
Size of cache to be used by osm2pgsql (in MB)

.TP
\fB\-\-socket\-timeout\fR \fI\,SOCKET_TIMEOUT\/\fR
Set timeout for file downloads

.SH OPTIONS 'nominatim special-phrases'
usage: nominatim special-phrases [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                                 [--import-from-wiki] [--import-from-csv FILE]
                                 [--no-replace]

    Import special phrases.
.br

.br
    Special phrases are search terms that narrow down the type of object
.br
    that should be searched. For example, you might want to search for
.br
    'Hotels in Barcelona'. The OSM wiki has a selection of special phrases
.br
    in many languages, which can be imported with this command.
.br

.br
    You can also provide your own phrases in a CSV file. The file needs to have
.br
    the following five columns:
.br
     * phrase \- the term expected for searching
.br
     * class \- the OSM tag key of the object type
.br
     * type \- the OSM tag value of the object type
.br
     * operator \- the kind of search to be done (one of: in, near, name, \-)
.br
     * plural \- whether the term is a plural or not (Y/N)
.br

.br
    An example file can be found in the Nominatim sources at
.br
    'test/testdb/full_en_phrases_test.csv'.
.br

.br
    The import can be further configured to ignore specific key/value pairs.
.br
    This is particularly useful when importing phrases from the wiki. The
.br
    default configuration excludes some very common tags like building=yes.
.br
    The configuration can be customized by putting a file `phrase\-settings.json`
.br
    with custom rules into the project directory or by using the `\-\-config`
.br
    option to point to another configuration file.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-import\-from\-wiki\fR
Import special phrases from the OSM wiki to the database

.TP
\fB\-\-import\-from\-csv\fR FILE
Import special phrases from a CSV file

.TP
\fB\-\-no\-replace\fR
Keep the old phrases and only add the new ones

.SH OPTIONS 'nominatim add-data'
usage: nominatim add-data [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                          (--file FILE | --diff FILE | --node ID | --way ID | --relation ID | --tiger-data DIR)
                          [--use-main-api] [--osm2pgsql-cache SIZE]
                          [--socket-timeout SOCKET_TIMEOUT]

    Add additional data from a file or an online source.
.br

.br
    This command allows to add or update the search data in the database.
.br
    The data can come either from an OSM file or single OSM objects can
.br
    directly be downloaded from the OSM API. This function only loads the
.br
    data into the database. Afterwards it still needs to be integrated
.br
    in the search index. Use the `nominatim index` command for that.
.br

.br
    The command can also be used to add external non\-OSM data to the
.br
    database. At the moment the only supported format is TIGER housenumber
.br
    data. See the online documentation at
.br
    https://nominatim.org/release\-docs/latest/customize/Tiger/
.br
    for more information.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-file\fR FILE
Import data from an OSM file or diff file

.TP
\fB\-\-diff\fR FILE
Import data from an OSM diff file (deprecated: use \-\-file)

.TP
\fB\-\-node\fR ID
Import a single node from the API

.TP
\fB\-\-way\fR ID
Import a single way from the API

.TP
\fB\-\-relation\fR ID
Import a single relation from the API

.TP
\fB\-\-tiger\-data\fR DIR
Add housenumbers from the US TIGER census database

.TP
\fB\-\-use\-main\-api\fR
Use OSM API instead of Overpass to download objects

.TP
\fB\-\-osm2pgsql\-cache\fR SIZE
Size of cache to be used by osm2pgsql (in MB)

.TP
\fB\-\-socket\-timeout\fR \fI\,SOCKET_TIMEOUT\/\fR
Set timeout for file downloads

.SH OPTIONS 'nominatim index'
usage: nominatim index [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                       [--boundaries-only] [--no-boundaries] [--minrank RANK]
                       [--maxrank RANK]

    Reindex all new and modified data.
.br

.br
    Indexing is the process of computing the address and search terms for
.br
    the places in the database. Every time data is added or changed, indexing
.br
    needs to be run. Imports and replication updates automatically take care
.br
    of indexing. For other cases, this function allows to run indexing manually.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-boundaries\-only\fR
Index only administrative boundaries.

.TP
\fB\-\-no\-boundaries\fR
Index everything except administrative boundaries.

.TP
\fB\-\-minrank\fR RANK, \fB\-r\fR RANK
Minimum/starting rank

.TP
\fB\-\-maxrank\fR RANK, \fB\-R\fR RANK
Maximum/finishing rank

.SH OPTIONS 'nominatim refresh'
usage: nominatim refresh [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                         [--postcodes] [--word-tokens] [--word-counts]
                         [--address-levels] [--functions] [--wiki-data]
                         [--secondary-importance] [--importance] [--website]
                         [--data-object OBJECT] [--data-area OBJECT]
                         [--no-diff-updates] [--enable-debug-statements]

    Recompute auxiliary data used by the indexing process.
.br

.br
    This sub\-commands updates various static data and functions in the database.
.br
    It usually needs to be run after changing various aspects of the
.br
    configuration. The configuration documentation will mention the exact
.br
    command to use in such case.
.br

.br
    Warning: the 'update' command must not be run in parallel with other update
.br
             commands like 'replication' or 'add\-data'.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-postcodes\fR
Update postcode centroid table

.TP
\fB\-\-word\-tokens\fR
Clean up search terms

.TP
\fB\-\-word\-counts\fR
Compute frequency of full\-word search terms

.TP
\fB\-\-address\-levels\fR
Reimport address level configuration

.TP
\fB\-\-functions\fR
Update the PL/pgSQL functions in the database

.TP
\fB\-\-wiki\-data\fR
Update Wikipedia/data importance numbers

.TP
\fB\-\-secondary\-importance\fR
Update secondary importance raster data

.TP
\fB\-\-importance\fR
Recompute place importances (expensive!)

.TP
\fB\-\-website\fR
Refresh the directory that serves the scripts for the web API

.TP
\fB\-\-data\-object\fR OBJECT
Mark the given OSM object as requiring an update (format: [NWR]<id>)

.TP
\fB\-\-data\-area\fR OBJECT
Mark the area around the given OSM object as requiring an update (format: [NWR]<id>)

.TP
\fB\-\-no\-diff\-updates\fR
Do not enable code for propagating updates

.TP
\fB\-\-enable\-debug\-statements\fR
Enable debug warning statements in functions

.SH OPTIONS 'nominatim admin'
usage: nominatim admin [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                       (--warm | --check-database | --migrate | --analyse-indexing | --collect-os-info | --clean-deleted AGE)
                       [--search-only] [--reverse-only]
                       [--osm-id OSM_ID | --place-id PLACE_ID]

    Analyse and maintain the database.
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-warm\fR
Warm database caches for search and reverse queries

.TP
\fB\-\-check\-database\fR
Check that the database is complete and operational

.TP
\fB\-\-migrate\fR
Migrate the database to a new software version

.TP
\fB\-\-analyse\-indexing\fR
Print performance analysis of the indexing process

.TP
\fB\-\-collect\-os\-info\fR
Generate a report about the host system information

.TP
\fB\-\-clean\-deleted\fR AGE
Clean up deleted relations

.TP
\fB\-\-search\-only\fR
Only pre\-warm tables for search queries

.TP
\fB\-\-reverse\-only\fR
Only pre\-warm tables for reverse queries

.TP
\fB\-\-osm\-id\fR \fI\,OSM_ID\/\fR
Analyse indexing of the given OSM object

.TP
\fB\-\-place\-id\fR \fI\,PLACE_ID\/\fR
Analyse indexing of the given Nominatim object

.SH OPTIONS 'nominatim export'
usage: nominatim export [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                        [--output-type {country,state,county,city,suburb,street,path}]
                        [--output-format OUTPUT_FORMAT] [--language LANGUAGE]
                        [--restrict-to-country COUNTRY_CODE]
                        [--restrict-to-osm-node ID] [--restrict-to-osm-way ID]
                        [--restrict-to-osm-relation ID]

    Export places as CSV file from the database.
.br

.br

.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-output\-type\fR {country,state,county,city,suburb,street,path}
Type of places to output (default: street)

.TP
\fB\-\-output\-format\fR \fI\,OUTPUT_FORMAT\/\fR
Semicolon\-separated list of address types (see \-\-output\-type). Additionally accepts:placeid,postcode

.TP
\fB\-\-language\fR \fI\,LANGUAGE\/\fR
Preferred language for output (use local name, if omitted)

.TP
\fB\-\-restrict\-to\-country\fR COUNTRY_CODE
Export only objects within country

.TP
\fB\-\-restrict\-to\-osm\-node\fR ID
Export only children of this OSM node

.TP
\fB\-\-restrict\-to\-osm\-way\fR ID
Export only children of this OSM way

.TP
\fB\-\-restrict\-to\-osm\-relation\fR ID
Export only children of this OSM relation

.SH OPTIONS 'nominatim convert'
usage: nominatim convert [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                         [--format {sqlite}] --output OUTPUT [--with-reverse]
                         [--with-search] [--with-details]

 Convert an existing database into a different format. (EXPERIMENTAL)
.br

.br
        Dump a read\-only version of the database in a different format.
.br
        At the moment only a SQLite database suitable for reverse lookup
.br
        can be created.
.br
    


.TP
\fB\-\-format\fR {sqlite}
Format of the output database (must be sqlite currently)

.TP
\fB\-\-output\fR \fI\,OUTPUT\/\fR, \fB\-o\fR \fI\,OUTPUT\/\fR
File to write the database to.

.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-with\-reverse\fR, \fB\-\-without\-reverse\fR
Enable/disable support for reverse and lookup API (default: enabled)

.TP
\fB\-\-with\-search\fR, \fB\-\-without\-search\fR
Enable/disable support for search API (default: disabled)

.TP
\fB\-\-with\-details\fR, \fB\-\-without\-details\fR
Enable/disable support for details API (default: enabled)

.SH OPTIONS 'nominatim serve'
usage: nominatim serve [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                       [--server SERVER] [--engine {php,falcon,starlette}]

    Start a simple web server for serving the API.
.br

.br
    This command starts a built\-in webserver to serve the website
.br
    from the current project directory. This webserver is only suitable
.br
    for testing and development. Do not use it in production setups!
.br

.br
    There are different webservers available. The default 'php' engine
.br
    runs the classic PHP frontend. The other engines are Python servers
.br
    which run the new Python frontend code. This is highly experimental
.br
    at the moment and may not include the full API.
.br

.br
    By the default, the webserver can be accessed at: http://127.0.0.1:8088
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-server\fR \fI\,SERVER\/\fR
The address the server will listen to.

.TP
\fB\-\-engine\fR {php,falcon,starlette}
Webserver framework to run. (default: falcon)

.SH OPTIONS 'nominatim search'
usage: nominatim search [-h] [-q] [-v] [--project-dir DIR] [-j NUM]
                        [--query QUERY] [--amenity AMENITY] [--street STREET]
                        [--city CITY] [--county COUNTY] [--state STATE]
                        [--country COUNTRY] [--postalcode POSTALCODE]
                        [--format {xml,geojson,geocodejson,json,jsonv2,debug}]
                        [--addressdetails] [--extratags] [--namedetails]
                        [--lang LANGS]
                        [--polygon-output {geojson,kml,svg,text}]
                        [--polygon-threshold TOLERANCE] [--countrycodes CC,..]
                        [--exclude_place_ids ID,..] [--limit LIMIT]
                        [--viewbox X1,Y1,X2,Y2] [--bounded] [--no-dedupe]

    Execute a search query.
.br

.br
    This command works exactly the same as if calling the /search endpoint on
.br
    the web API. See the online documentation for more details on the
.br
    various parameters:
.br
    https://nominatim.org/release\-docs/latest/api/Search/
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-query\fR \fI\,QUERY\/\fR
Free\-form query string

.TP
\fB\-\-amenity\fR \fI\,AMENITY\/\fR
Structured query: name and/or type of POI

.TP
\fB\-\-street\fR \fI\,STREET\/\fR
Structured query: housenumber and street

.TP
\fB\-\-city\fR \fI\,CITY\/\fR
Structured query: city, town or village

.TP
\fB\-\-county\fR \fI\,COUNTY\/\fR
Structured query: county

.TP
\fB\-\-state\fR \fI\,STATE\/\fR
Structured query: state

.TP
\fB\-\-country\fR \fI\,COUNTRY\/\fR
Structured query: country

.TP
\fB\-\-postalcode\fR \fI\,POSTALCODE\/\fR
Structured query: postcode

.TP
\fB\-\-format\fR {xml,geojson,geocodejson,json,jsonv2,debug}
Format of result

.TP
\fB\-\-addressdetails\fR
Include a breakdown of the address into elements

.TP
\fB\-\-extratags\fR
Include additional information if available (e.g. wikipedia link, opening hours)

.TP
\fB\-\-namedetails\fR
Include a list of alternative names

.TP
\fB\-\-lang\fR LANGS, \fB\-\-accept\-language\fR LANGS
Preferred language order for presenting search results

.TP
\fB\-\-polygon\-output\fR {geojson,kml,svg,text}
Output geometry of results as a GeoJSON, KML, SVG or WKT

.TP
\fB\-\-polygon\-threshold\fR TOLERANCE
Simplify output geometry.Parameter is difference tolerance in degrees.

.TP
\fB\-\-countrycodes\fR CC,..
Limit search results to one or more countries

.TP
\fB\-\-exclude_place_ids\fR ID,..
List of search object to be excluded. Each object can be a Nominatim \fBplace_id\fR or an OSM object in the form \fB<osm_type><osm_id>\fR

.TP
\fB\-\-limit\fR \fI\,LIMIT\/\fR
Limit the number of returned results

.TP
\fB\-\-viewbox\fR X1,Y1,X2,Y2
Preferred area to find search results

.TP
\fB\-\-bounded\fR
Strictly restrict results to viewbox area

.TP
\fB\-\-no\-dedupe\fR
Do not remove duplicates from the result list

.SH OPTIONS 'nominatim reverse'
usage: nominatim reverse [-h] [-q] [-v] [--project-dir DIR] [-j NUM] --lat LAT
                         --lon LON [--zoom ZOOM] [--layer LAYER]
                         [--format {xml,geojson,geocodejson,json,jsonv2,debug}]
                         [--addressdetails] [--extratags] [--namedetails]
                         [--lang LANGS]
                         [--polygon-output {geojson,kml,svg,text}]
                         [--polygon-threshold TOLERANCE]

    Execute API reverse query.
.br

.br
    This command works exactly the same as if calling the /reverse endpoint on
.br
    the web API. See the online documentation for more details on the
.br
    various parameters:
.br
    https://nominatim.org/release\-docs/latest/api/Reverse/
.br
    



.TP
\fB\-q\fR, \fB\-\-quiet\fR
Print only error messages

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Increase verboseness of output

.TP
\fB\-\-project\-dir\fR DIR
Base directory of the Nominatim installation (default:.)

.TP
\fB\-j\fR NUM, \fB\-\-threads\fR NUM
Number of parallel threads to use

.TP
\fB\-\-lat\fR \fI\,LAT\/\fR
Latitude of coordinate to look up (in WGS84)

.TP
\fB\-\-lon\fR \fI\,LON\/\fR
Longitude of coordinate to look up (in WGS84)

.TP
\fB\-\-zoom\fR \fI\,ZOOM\/\fR
Level of detail required for the address

.TP
\fB\-\-layer\fR LAYER
OSM id to lookup in format <NRW><id> (may be repeated)

.TP
\fB\-\-format\fR {xml,geojson,geocodejson,json,jsonv2,debug}
Format of result

.TP
\fB\-\-addressdetails\fR
Include a breakdown of the address into elements

.TP
\fB\-\-extratags\fR

ChangeLog
5.2.0
 * increase minimum required Python to 3.9
 * index and output entrances of buildings and areas (thanks @emlove)
 * name tags used for creating display names are now configurable
   (thanks @astridx)
 * new pattern-replacement query preprocessor (thanks @TuringVerified)
 * special phrases can now be filtered by presence of tags (thanks @anqixxx)
 * lua import style now always includes tags required by Nominatim
 * improved query time reporting and logging
 * improve word matching for languages with no word boundaries
 * POIs with addresses inherited from surrounding building are no
   longer returned in the address layer
 * avoid creating a directory for the tokenizer when not needed
 * replace behave with pytest-bdd for BDD testing
 * refactoring and performance improvements to query parsing
 * various smaller updates to styles
 * remove English as default language for South Korea
 * remove Japanese word variants
 * updated country names for Norwegians (thanks @Johannes-Andersen)
 * remove support for deprecated osm2pgsql gazetteer style
 * fix updating of importances (also needs to update search_name table)
 * fix query for deletable endpoint to use index again
 * fix reindexing of contained places when a boundary is deleted and reinstated
 * fix difference computation error when updating postcodes
 * bracket handling sanitizer no longer strips bracket terms in the middle of
   name
 * reduce precision of stored coordinates to 7-digits everywhere
 * avoid ST_Relate as it seems buggy on some systems
 * remove setting for logging queries in DB, no longer functional
 * postcode updates no longer require a project directory (needed for tests)
 * refactor locale handling code (thanks @anqixxx)
 * code updates for newer Python (thanks @emmanuel-ferdman)
 * better test coverage (thanks @asharmalik19)
 * various fixes and improvements to documentation
   (thanks @anqixxx, @dave-meyer, @hasandiwan)

5.1.0
 * replace datrie with simple internal trie implementation
 * add pattern-based postcode parser for queries,
   postcodes no longer need to be present in OSM to be found
 * take variants into account when computing token similarity
 * add extratags output to geocodejson format
 * fix default layer setting used for structured queries
 * update abbreviation lists for Russian and English
   (thanks @shoorick, @IvanShift, @mhsrn21)
 * fix variant generation for Norwegian
 * fix normalization around space-like characters
 * improve postcode search and handling of postcodes in queries
 * reorganise internal query structure and get rid of slow enums
 * enable code linting for tests
 * various code moderinsations in test code (thanks @eumiro)
 * remove setting osm2pgsql location via config.lib_dir
 * make SQL functions parallel save as far as possible (thanks @otbutz)
 * various fixes and improvements to documentation (thanks @TuringVerified)

5.0.0
 * increase required versions for PostgreSQL (12+), PostGIS (3.0+)
 * remove installation via cmake and debundle osm2pgsql
 * remove deprecated PHP frontend
 * remove deprecated legacy tokenizer
 * add configurable pre-processing of queries
 * add query pre-processor to split up Japanese addresses
 * rewrite of osm2pgsql style implementation
   (also adds support for osm2pgsql-themepark)
 * reduce the number of SQL queries needed to complete a 'lookup' call
 * improve computation of centroid for lines with only two points
 * improve bbox output for postcode areas
 * improve result order by returning the largest object when other things are
   equal
 * add fallback for reverse geocoding to default country tables
 * exclude postcode areas from reverse geocoding
 * disable search endpoint when database is reverse-only (regression)
 * minor performance improvements to area split algorithm
 * switch table and index creation to use autocommit mode to avoid deadlocks
 * drop overly long ways during import
 * restrict automatic migrations to versions 4.3+
 * switch linting from pylint to flake8
 * switch tests to use a wikimedia test file in the new CSV style
 * various fixes and improvements to documentation

4.5.0
 * allow building Nominatim as a pip package
 * make osm2pgsql building optional
 * switch importer to psycopg3
 * allow output format of web search to be customized in self-installations
 * look up potential postcode areas for postcode results
 * add word usage statistics for address terms
 * implement more light-weight CSV format for wiki importance tables
 * rewrite SQL for place search to use window functions
 * increase search radius when filtering by postcode
 * prefer POI points over POI areas
 * reintroduce full terms for address terms in search_name table
 * reindex postcodes when their parent is deleted
 * indexing: precompute counts of affected rows
 * ensure consistent country assignments for overlapping countries
 * make Nominatim[Async]API context manager to ensure proper calling of
   close()
 * make usage of project dir optional for library
 * drop interpolations when no parent can be found
 * style tweaks to reflect OSM usage (man_made, highway and others)
 * deprecation of: bundled osm2pgsql, legacy tokenizer, PHP frontend
 * make documentation buildable without CMake
 * various fixes and improvements to documentation

4.4.1
 * fix geocodejson output: admin level output should only print boundaries
 * updating: restrict invalidation of child objects on large street features
 * restrict valid interpolation house numbers to 0-999999
 * fix import error when SQLAlchemy 1.4 and psycopg3 are installed
 * various typo fixes in the documentation

4.4.0
 * add export to SQLite database and SQLite support for the frontend
 * switch to Python frontend as the default frontend
 * update to osm2pgsql 1.11.0
 * add support for new osm2pgsql middle table format
 * simplify geometry for large polygon objects not used in addresses
 * various performance tweaks for search in Python frontend

nominatim/nominatim.py
#! /usr/bin/env python3
#-----------------------------------------------------------------------------
# nominatim - [description]
#-----------------------------------------------------------------------------
#
# Indexing tool for the Nominatim database.
#
# Based on C version by Brian Quinion
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#-----------------------------------------------------------------------------
from argparse import ArgumentParser, RawDescriptionHelpFormatter
import logging
import sys
import getpass

from indexer.indexer import Indexer

def nominatim_arg_parser():
    """ Setup the command-line parser for the tool.
    """
    parser = ArgumentParser(description="Indexing tool for Nominatim.",
                            formatter_class=RawDescriptionHelpFormatter)

    parser.add_argument('-d', '--database',
                        dest='dbname', action='store', default='nominatim',
                        help='Name of the PostgreSQL database to connect to.')
    parser.add_argument('-U', '--username',
                        dest='user', action='store',
                        help='PostgreSQL user name.')
    parser.add_argument('-W', '--password',
                        dest='password_prompt', action='store_true',
                        help='Force password prompt.')
    parser.add_argument('-H', '--host',
                        dest='host', action='store',
                        help='PostgreSQL server hostname or socket location.')
    parser.add_argument('-P', '--port',
                        dest='port', action='store',
                        help='PostgreSQL server port')
    parser.add_argument('-b', '--boundary-only',
                        dest='boundary_only', action='store_true',
                        help='Only index administrative boundaries (ignores min/maxrank).')
    parser.add_argument('-r', '--minrank',
                        dest='minrank', type=int, metavar='RANK', default=0,
                        help='Minimum/starting rank.')
    parser.add_argument('-R', '--maxrank',
                        dest='maxrank', type=int, metavar='RANK', default=30,
                        help='Maximum/finishing rank.')
    parser.add_argument('-t', '--threads',
                        dest='threads', type=int, metavar='NUM', default=1,
                        help='Number of threads to create for indexing.')
    parser.add_argument('-v', '--verbose',
                        dest='loglevel', action='count', default=0,
                        help='Increase verbosity')

    return parser

if __name__ == '__main__':
    OPTIONS = nominatim_arg_parser().parse_args(sys.argv[1:])

    logging.basicConfig(stream=sys.stderr, format='%(levelname)s: %(message)s',
                        level=max(3 - OPTIONS.loglevel, 0) * 10)

    OPTIONS.password = None
    if OPTIONS.password_prompt:
        PASSWORD = getpass.getpass("Database password: ")
        OPTIONS.password = PASSWORD

    if OPTIONS.boundary_only:
        Indexer(OPTIONS).index_boundaries()
    else:
        Indexer(OPTIONS).index_by_rank()
